Source/JavaScriptCore/assembler/AssemblerCommon.h

```diff
@@ -379,32 +390,50 @@ static ALWAYS_INLINE void* memcpyAtomicIfPossible(void* dst, const void* src, si
     default:
         break;
     }
-#endif
+    RELEASE_ASSERT_NOT_REACHED();
+    return nullptr;
+}
+
+ALWAYS_INLINE void* memcpyTearing(void* dst, const void* src, size_t n)
+{
+    // We should expect these instructions to be torn, so let's verify that.
+    if (Options::fuzzAtomicJITMemcpy()) [[unlikely]] {
+        auto* d = reinterpret_cast<uint8_t*>(dst);
+        auto* s = reinterpret_cast<const uint8_t*>(src);
+        for (size_t i = 0; i < n; ++i, ++s, ++d) {
+            *d = *s;
+            WTF::storeLoadFence();
+        }
+    }
     return memcpy(dst, src, n);
 }
 
-static void* performJITMemcpy(void* dst, const void* src, size_t n);
+static ALWAYS_INLINE void* memcpyAtomicIfPossible(void* dst, const void* src, size_t n)
+{
+    if (n == 1 || n == 2 || n == 4 || (!is32Bit() && n == 8))
+        return memcpyAtomic(dst, src, n);
+    return memcpyTearing(dst, src, n);
+}
+
+template<RepatchingInfo repatch>
+void* performJITMemcpy(void* dst, const void* src, size_t n);
 
-template<MachineCodeCopyMode copy>
+template<RepatchingInfo repatch>
 ALWAYS_INLINE void* machineCodeCopy(void* dst, const void* src, size_t n)
 {
-#if CPU(ARM_THUMB2)
-    // For thumb instructions, we want to avoid the case where we have
-    // to repatch a 32-bit instruction that crosses 2 words.
-    bool isAligned = (dst == WTF::roundUpToMultipleOf<4>(dst));
-    if (n == 2 * sizeof(int16_t) && isAligned) {
-        *static_cast<uint32_t*>(dst) = *static_cast<const uint32_t*>(src);
-        return dst;
-    }
-    if (n == 1 * sizeof(int16_t)) {
-        *static_cast<uint16_t*>(dst) = *static_cast<const uint16_t*>(src);
-        return dst;
+    static_assert(!(*repatch).contains(RepatchingFlag::Flush));
+    if constexpr (is32Bit()) {
+        // Avoid unaligned accesses.
+        if (WTF::isAligned(dst, n))
+            return memcpyAtomicIfPossible(dst, src, n);
+        return memcpyTearing(dst, src, n);
```

Comment:
I'm guessing this doesn't look at `RepatchingFlag::Atomic` on 32 bits to reproduce the existing behavior? If so, wouldn't mind a comment explaining that :innocent: